analysis/src/diff.rs

```diff
@@ -178,16 +188,58 @@ impl DiffAnalyzer {
 
         Ok(Repository::open(path)?)
     }
+
+    fn setup_remote(&self, repo: &Repository, url: &str) -> Result<()> {
+        let remote_name = "source";
+        let mut remote = repo.remote(remote_name, url)?;
+        remote.connect(Direction::Fetch)?;
+        let default = remote.default_branch()?;
+        let default = default.as_str().unwrap();
+        remote.fetch(&[default], None, None)?;
+        Ok(())
+    }
+
+    fn analyze_diff<'a>(
+        &'a self,
+        repo: &'a Repository,
+        commit_a: &Commit,
+        commit_b: &Commit,
+    ) -> Result<Diff<'a>> {
+        let tree_a = commit_a.tree()?;
+        let tree_b = commit_b.tree()?;
+        let mut opts = DiffOptions::new();
+        let diff = repo.diff_tree_to_tree(Some(&tree_a), Some(&tree_b), Some(&mut opts))?;
+
+        //diff.print(DiffFormat::Patch, mut cb: F)
+        let stats = diff.stats()?;
+        let mut format = git2::DiffStatsFormat::NONE;
+        format |= git2::DiffStatsFormat::FULL;
+        let buf = stats.to_buf(format, 80)?;
+        print!(
+            "difference between crates.io and source is:\n {}",
+            std::str::from_utf8(&*buf).unwrap()
+        );
+
+        Ok(diff)
+    }
```

Comment:
One potential issue with this way of diffing crates is that it probably doesn't work 100% correctly with crates that are a part of a workspace. Eg you should take a look at running this on the guppy crate since that repo has a handful of crates in a workspace.

Reply:
I thought about those cases. My thinking was that in those cases the diff would be files that are in git source and not in crates.io such as the files in other crates of the workspace. Now for crate_source_diff analysis, we are already ignoring all files that are in git source and not in crates.io as that could arise for various reasons like this. 

However, if there are files in crates.io but not in git source and files in both places but different, those are the red flags. On that heuristic, this analysis is simple yet effective. That's my rationale here. Let me know if you have any suggestion.

Reply:
So the issue I'm worried about is the differing directory level, for instance:

the `guppy` crate lives in the repo `https://github.com/facebookincubator/cargo-guppy` but it doesn't live at the root of the repository, its located at `/guppy` where as when you're looking at the code from the crate you've downloaded from crates.io, that code will be located at the root of the project `/`. So in that instance wouldn't all the files be different?

I'm really just theorizing here and perhaps `git2`'s diff functionality is smart enough to do the rename detection when diffing. Would you be able to add a test for this particular instance so that we can be confident your solution covers this case?

Reply:
Oh that's a good catch. Yes, I didn't think about the path difference issue. Let me check.